lib/useQuery.js

```diff
@@ -137,16 +147,24 @@ export function useQuery(serviceName, options = {}, queryHookOptions = {}) {
     updateCache,
     isPending,
     isCacheSufficient,
+    state.dirty,
   ])
 
-  const refetch = useCallback(() => dispatch({ type: 'reset' }), [dispatch])
-
-  // refetch if the query changes
+  // refetch if queryId changed (which means the query changed),
+  // but not if this is the first render
+  const previousQueryIdRef = useRef(null)
   useEffect(() => {
-    if (!isCacheSufficient) {
-      refetch()
+    if (
+      previousQueryIdRef.current &&
+      previousQueryIdRef.current !== queryId &&
+      !isCacheSufficient
+    ) {
+      dispatch({ type: 'reset' })
     }
-  }, [refetch, queryId, isCacheSufficient])
+    previousQueryIdRef.current = queryId
+  }, [dispatch, queryId, isCacheSufficient])
+
+  const refetch = useCallback(() => dispatch({ type: 'reset' }), [dispatch])
```

Comment:
I removed the introduction of initialisedRef, because I do not think it was modelling the behaviour correctly. We only marked initialisedRef to true after the fetch promise completed, but the query could have changed in the meantime and we would never refetch - that could introduce subtle bugs.

With the previousQueryIdRef initialised to null, we solve the same issue of avoiding fetching on initial render - but we do initiate refetch when query changes.